pkg/controller/jobs/statefulset/statefulset_reconciler.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package statefulset
+
+import (
+	"context"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+	"sigs.k8s.io/kueue/pkg/controller/jobs/pod"
+	clientutil "sigs.k8s.io/kueue/pkg/util/client"
+	"sigs.k8s.io/kueue/pkg/util/parallelize"
+)
+
+// +kubebuilder:rbac:groups="apps",resources=statefulsets,verbs=get;list;watch
+
+var (
+	_ jobframework.JobReconcilerInterface = (*Reconciler)(nil)
+)
+
+type Reconciler struct {
+	client client.Client
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	sts := &appsv1.StatefulSet{}
+	err := r.client.Get(ctx, req.NamespacedName, sts)
+	if err != nil {
+		// we'll ignore not-found errors, since there is nothing to do.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("statefulset", klog.KObj(sts))
+	ctx = ctrl.LoggerInto(ctx, log)
+	log.V(2).Info("Reconciling StatefulSet")
+
+	// For now, handle only scaling down to zero.
+	if ptr.Deref(sts.Spec.Replicas, 1) != 0 {
+		return ctrl.Result{}, nil
+	}
+
+	err = r.fetchAndFinalizePods(ctx, req.Namespace, req.Name)
```

Comment:
This lgtm, but I'm wondering if there is a simpler solution.

Maybe we don't need to add the finalizers for owners managed by Kueue in the first place (IsOwnerManagedByKueue function) .

IIUC the purpose of the finalizers was to allow for running standalone pods or pod groups (without owner), but if we have StatefulSet as owner, the finalizer seems just complicate the flow.

If this is possible we would not need the Reconciler at all - neither now, nor for the arbitrary scale downs.

@mbobrovskyi could you investigate this possibility and see if we hit any problems? You may explore this in a separate PR so that we don't override this potential solution.

Reply:
I synced with @mbobrovskyi on that and we still need finalizers probably to support StatefulSet eviction. So, when evicting we want to keep the workload, because only the workload has the information that the StatefulSet is suspended. So, if we deleted all pods the workload would be gone, and thus the StatefulSet would "unsuspend itself. So, I think the current implementation is ok.